### PR TITLE
Expose SMTP response and response code to Objective-C API

### PR DESCRIPTION
**Pull request by kolyuchiy**

### CODE REVIEW

File: src/objc/smtp/MCOSMTPSendOperation.mm [hunk 1]
```diff
@@ -99,7 +101,17 @@ - (void)operationCompleted
     if (_completionBlock == NULL)
         return;
     
-    NSError * error = [NSError mco_errorWithErrorCode:MCO_NATIVE_INSTANCE->error()];
+    NSMutableDictionary *userInfo = [NSMutableDictionary dictionary];
```
Comment (dinhviethoa) [inline src/objc/smtp/MCOSMTPSendOperation.mm]: Could you take a look at how it's done in MCOIMAPCheckAccountOperation.mm?

File: src/objc/smtp/MCOSMTPSendOperation.mm [hunk 2]
```diff
@@ -99,7 +101,17 @@ - (void)operationCompleted
...
+    MCOSMTPSession *session = [self session];
+    if (session.lastSMTPResponse) {
```
Comment (dinhviethoa) [inline src/objc/smtp/MCOSMTPSendOperation.mm]: `[session lastSMTPResponse]` for consistency.

File: src/objc/smtp/MCOSMTPSession.mm [hunk 3]
```diff
@@ -98,6 +98,14 @@ - (void)dealloc {
 MCO_OBJC_SYNTHESIZE_BOOL(setUseHeloIPEnabled, useHeloIPEnabled)
 MCO_OBJC_SYNTHESIZE_SCALAR(dispatch_queue_t, dispatch_queue_t, setDispatchQueue, dispatchQueue);
 
+- (NSString *)lastSMTPResponse {
+    return [NSObject mco_objectWithMCObject:MCO_NATIVE_INSTANCE->lastSMTPResponse()];
+}
+
+- (NSInteger)lastSMTPResponseCode {
```
Comment (dinhviethoa) [inline src/objc/smtp/MCOSMTPSession.mm]: use int instead of NSInteger.

File: src/objc/smtp/MCOSMTPOperation.h [hunk 4]
```diff
@@ -14,6 +14,9 @@
 
 /** This is an asynchronous SMTP operation, used for sending messages. */
 
+extern NSString * _Nonnull const kMCOSMTPOperationResponseKey;
```
Comment (dinhviethoa) [inline src/objc/smtp/MCOSMTPOperation.h]: Could you use:

```
#define MCOSMTPResponseKey @"MCOSMTPResponseKey"
#define MCOSMTPResponseCodeKey @"MCOSMTPResponseCodeKey"
```

And have them in MCOConstant.h instead?

File: src/objc/smtp/MCOSMTPSendOperation.mm (hunk 1 above)
Comment (kolyuchiy) [inline src/objc/smtp/MCOSMTPSendOperation.mm]: done

File: src/async/smtp/MCSMTPAsyncSession.cpp [hunk 5]
```diff
@@ -184,6 +184,15 @@ bool SMTPAsyncSession::useHeloIPEnabled()
     return mSession->useHeloIPEnabled();
 }
 
+String * SMTPAsyncSession::lastSMTPResponse()
```
Comment (dinhviethoa) [inline src/async/smtp/MCSMTPAsyncSession.cpp]: Please don't expose them here. There might be some thread safety issue.

File: src/async/smtp/MCSMTPAsyncSession.h [hunk 6]
```diff
@@ -51,6 +51,10 @@ namespace mailcore {
         virtual void setUseHeloIPEnabled(bool enabled);
         virtual bool useHeloIPEnabled();
         
+        virtual String * lastSMTPResponse();
```
Comment (dinhviethoa) [inline src/async/smtp/MCSMTPAsyncSession.h]: Remove them from the API.

File: src/core/smtp/MCSMTPSession.h [hunk 7]
```diff
@@ -49,6 +49,10 @@ namespace mailcore {
         
         virtual void setUseHeloIPEnabled(bool enabled);
         virtual bool useHeloIPEnabled();
+
+        virtual String * lastSMTPResponse();
```
Comment (dinhviethoa) [inline src/core/smtp/MCSMTPSession.h]: They're good here.

File: src/objc/smtp/MCOSMTPLoginOperation.mm [hunk 8]
```diff
@@ -56,8 +58,24 @@ - (void)operationCompleted
     if (_completionBlock == NULL)
         return;
     
-    NSError * error = [NSError mco_errorWithErrorCode:MCO_NATIVE_INSTANCE->error()];
-    _completionBlock(error);
+    nativeType *op = MCO_NATIVE_INSTANCE;
```
Comment (dinhviethoa) [inline src/objc/smtp/MCOSMTPLoginOperation.mm]: Could you actually create in MCOSMTPOperation+Private.h, a method:

```
- (NSError *) error
{
    NSError * error = [NSError mco_errorWithErrorCode:MCO_NATIVE_INSTANCE->error()];
   if (error != nil) {
        MCOSMTPSession *session = [self session];
        if (nativeOp->lastSMTPResponse() || nativeOp->lastSMTPResponseCode()) {
            NSMutableDictionary * userInfo = [[error userInfo] mutableCopy];
            if (nativeOp->lastSMTPResponse()) {
                userInfo[MCOSMTPResponseKey] = MCO_TO_OBJC(nativeOp->lastSMTPResponse());
            }
            if (nativeOp->lastSMTPResponseCode()) {
                userInfo[MCOSMTPResponseCodeKey] = @(session nativeOp());
            }
            error = [NSError errorWithDomain:[error domain] code:[error code] userInfo:userInfo];
        }
   }
  return error;
}
```

And use that method everywhere that code is duplicated.

File: src/objc/smtp/MCOSMTPSession.h [hunk 9]
```diff
@@ -67,6 +67,10 @@
 */
 @property (nonatomic, assign, getter=isUseHeloIPEnabled) BOOL useHeloIPEnabled;
 
+@property (nonatomic, copy, readonly) NSString *lastSMTPResponse;
```
Comment (dinhviethoa) [inline src/objc/smtp/MCOSMTPSession.h]: Don't expose them here.

File: src/objc/smtp/MCOSMTPSession.mm [hunk 10]
```diff
@@ -98,6 +98,14 @@ - (void)dealloc {
 MCO_OBJC_SYNTHESIZE_BOOL(setUseHeloIPEnabled, useHeloIPEnabled)
 MCO_OBJC_SYNTHESIZE_SCALAR(dispatch_queue_t, dispatch_queue_t, setDispatchQueue, dispatchQueue);
 
+- (NSString *)lastSMTPResponse {
```
Comment (dinhviethoa) [inline src/objc/smtp/MCOSMTPSession.mm]: Remove the implementation here.

**Comment by dinhviethoa** [comment]
Could you expose lastSMTPResponse and lastSMTPResponseCode in SMTPOperation instead?
It will lead to ensuring thread safety.
I added some comments to guide you through the refactoring.
Thanks!

File: src/objc/smtp/MCOSMTPOperation+Private.h [hunk 11]
```diff
@@ -16,6 +16,8 @@
 
 @property (nonatomic, retain) MCOSMTPSession * session;
 
+@property (nonatomic, readonly, copy) NSError * error;
```
Comment (dinhviethoa) [inline src/objc/smtp/MCOSMTPOperation+Private.h]: Could you replace with:

```
- (NSError *) _errorFromNativeOperation;
```

File: src/objc/smtp/MCOSMTPOperation.mm [hunk 12]
```diff
@@ -56,4 +56,22 @@ - (MCOSMTPSession *) session
     return _session;
 }
 
+- (NSError *)error {
```
Comment (dinhviethoa) [inline src/objc/smtp/MCOSMTPOperation.mm]: _errorFromNativeOperation

**Comment by dinhviethoa** [comment]
There's one last change before I merge this pull request.
Thanks for adding that feature!

**Pull request merged**